src/liballoc/collections/vec_deque.rs

```diff
@@ -1834,8 +1850,17 @@ impl<T> VecDeque<T> {
     #[inline]
     #[stable(feature = "append", since = "1.4.0")]
     pub fn append(&mut self, other: &mut Self) {
-        // naive impl
-        self.extend(other.drain(..));
+        // Guarantees there is space in `self` for `other
+        self.reserve(other.len());
+
+        unsafe {
+            let (src_high, src_low) = other.as_slices();
+            self.copy_slice(src_low);
+            self.copy_slice(src_high);
```

Comment:
If the second `copy_slice` call panics after the first one succeeded, then some elements will be dropped twice: once when `self` is dropped, because `self.head` was already updated, and once when `other` is dropped, because `other.tail` has not been updated. 

That is, whether this is safe depends on whether `copy_slice` can panic. I think it is important to:

* state here that this snippet is only safe when `copy_slice` cannot panic,
* document in the `copy_slice` docs that the implementation of the function never panics when all pre-conditions are upheld by the caller
* triple-check that `copy_slice` never panics

Also, the following is not a strict requirement, but more like a `std` lib convention: the `unsafe` blocks in the std library tend to cover the whole function instead of the minimal amount of code required to make it compile. Here, the `reserve` call is required to uphold `copy_slice` pre-conditions, and the `other.tail = other.head` statement is required to avoid double-dropping elements, so I also feel it is logical to put the whole `append` implementation inside an `unsafe` block.